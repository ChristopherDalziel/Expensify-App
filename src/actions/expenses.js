import { v4 as uuidv4 } from "uuid";
import database from "../firebase/firebase";

export const addExpense = (expense) => ({
  type: "ADD_EXPENSE",
  expense,
});

export const startAddExpense = (expenseData = {}) => {
  // returning a function, that works only because we are using thunk
  return (dispatch) => {
    // Setting up the default object
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0,
    } = expenseData;
    // Destructor our data
    const expense = { description, note, amount, createdAt };
    // Save our data
    database
      .ref("expenses")
      .push(expense)
      .then((ref) => {
        dispatch(
          addExpense({
            id: ref.key,
            ...expense,
          })
        );
      });
  };
};

export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id,
});

export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates,
});
