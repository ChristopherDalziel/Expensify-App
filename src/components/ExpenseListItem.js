import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";
import numeral from "numeral";
import styled from "@emotion/styled";

const ExpenseListItemContainer = styled.div`
  margin: 10px;
`;

const ExpenseListItem = ({ description, amount, createdAt, id }) => (
  <ExpenseListItemContainer>
    <Link to={`/edit/${id}`}>
      <h3>{description}</h3>
    </Link>
    <p>
      {numeral(amount / 100).format("$0,0.00")} -
      {moment(createdAt).format("MMMM Do, YYYY")}
    </p>
  </ExpenseListItemContainer>
);

export default ExpenseListItem;
