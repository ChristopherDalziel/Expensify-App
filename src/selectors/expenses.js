import moment from "moment";

const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter((expense) => {
      // We are now using moment dates here to align with the component that is using date picker rather than manually inputted dates
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment, "day")
        : true;

      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment, "day")
        : true;

      const matchText = expense.description
        .toLowerCase()
        .includes(text.toLowerCase());

      return startDateMatch && endDateMatch && matchText;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return a.createdAt < b.createdAt ? 1 : -1;
      }
      if (sortBy === "amount") {
        return a.amount < b.amount ? 1 : -1;
      }
    });
};

export default getVisibleExpenses;
