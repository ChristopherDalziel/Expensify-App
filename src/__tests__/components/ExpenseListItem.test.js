import React from "react";
import { shallow } from "enzyme";
import ExpenseListItem from "../../components/ExpenseListItem";
import expenses from "../fixtures/expenses";

import toJSON from "enzyme-to-json";

test("Should render expense list item with data", () => {
  const wrapper = shallow(
    <ExpenseListItem
      description={expenses[0].description}
      amount={expenses[0].amount}
      createdAt={expenses[0].createdAt}
      id={expenses[0].id}
    />
  );
  expect(toJSON(wrapper)).toMatchSnapshot();
});
