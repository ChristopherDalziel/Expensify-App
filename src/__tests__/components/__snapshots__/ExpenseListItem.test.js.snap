// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render expense list item with data 1`] = `
<Styled(div)>
  <Link
    to="/edit/1"
  >
    <h3>
      gum
    </h3>
  </Link>
  <p>
    $1.95
     -
    January 1st, 1970
  </p>
</Styled(div)>
`;
